posthog/hogql/property.py

```diff
@@ -512,45 +606,38 @@ def entity_to_expr(entity: RetentionEntity) -> ast.Expr:
     )
 
 
-def element_chain_key_filter(key: str, text: str, operator: PropertyOperator):
-    escaped = text.replace('"', r"\"")
-    if operator == PropertyOperator.IS_SET or operator == PropertyOperator.IS_NOT_SET:
-        value = r'[^"]+'
-    elif operator == PropertyOperator.ICONTAINS or operator == PropertyOperator.NOT_ICONTAINS:
-        value = rf'[^"]*{re.escape(escaped)}[^"]*'
-    elif operator == PropertyOperator.REGEX or operator == PropertyOperator.NOT_REGEX:
-        value = escaped
-    elif operator == PropertyOperator.EXACT or operator == PropertyOperator.IS_NOT:
-        value = re.escape(escaped)
-    else:
-        raise NotImplementedError(f"element_href_to_expr not implemented for operator {operator}")
-
-    regex = f'({key}="{value}")'
-    if operator == PropertyOperator.ICONTAINS or operator == PropertyOperator.NOT_ICONTAINS:
-        expr = parse_expr("elements_chain =~* {regex}", {"regex": ast.Constant(value=str(regex))})
-    else:
-        expr = parse_expr("elements_chain =~ {regex}", {"regex": ast.Constant(value=str(regex))})
-
-    if (
-        operator == PropertyOperator.IS_NOT_SET
-        or operator == PropertyOperator.NOT_ICONTAINS
-        or operator == PropertyOperator.IS_NOT
-        or operator == PropertyOperator.NOT_REGEX
-    ):
-        expr = ast.Call(name="not", args=[expr])
-    return expr
-
-
 def tag_name_to_expr(tag_name: str):
     regex = rf"(^|;){tag_name}(\.|$|;|:)"
     expr = parse_expr("elements_chain =~ {regex}", {"regex": ast.Constant(value=str(regex))})
     return expr
 
 
-def selector_to_expr(selector: str):
-    regex = build_selector_regex(Selector(selector, escape_slashes=False))
-    expr = parse_expr("elements_chain =~ {regex}", {"regex": ast.Constant(value=regex)})
-    return expr
+def selector_to_expr(selector_string: str):
+    selector = Selector(selector_string, escape_slashes=False)
+    exprs = []
+    regex = build_selector_regex(selector)
+    exprs.append(parse_expr("elements_chain =~ {regex}", {"regex": ast.Constant(value=regex)}))
```

Comment:
_Re-posting from Slack:_
So the only thing I'm wondering about is: we'll always have the full `elements_chain =~` _and_ the lightweight `elements_chain_{foo}` conditions all in the same `AND` expression. What's the guarantee that makes ClickHouse always execute the heavy `elements_chain` check last?